esmvalcore/_data_finder.py

```diff
@@ -33,29 +33,58 @@ def find_files(dirnames, filenames):
 
 
 def get_start_end_year(filename):
-    """Get the start and end year from a file name."""
+    """Get the start and end year from a file name.
+
+    Examples of allowed dates : 1980, 198001, 19801231,
+    1980123123, 19801231T23, 19801231T2359, 19801231T235959,
+    19801231T235959Z
+
+    Dates must be surrounded by - or _ or string start or string end
+    (after removing filename suffix)
+
+    Look first for two dates separated by - or _, then for one single
+    date, and if they are multiple, for one date at start or end
+    """
     stem = Path(filename).stem
     start_year = end_year = None
 
-    # First check for a block of two potential dates separated by _ or -
-    daterange = re.findall(r'([0-9]{4,12}[-_][0-9]{4,12})', stem)
+    date_pattern = r"(?P<year>[0-9]{4})(?P<month>[01][0-9]" + \
+                   "(?P<day>[0-3][0-9](T?(?P<hour>[0-2][0-9]" + \
+                   "(?P<minute>[0-5][0-9](?P<second>[0-5][0-9]Z?)?)?)?)?)?)?"
```

Comment:
Nice. Perhaps it's a bit more readable to first define a time pattern like
```
    time_pattern = r"(?P<hour>[0-2][0-9](?P<minute>[0-5][0-9](?P<second>[0-5][0-9])?)?Z?)"
    datetime_pattern = (r"(?P<year>[0-9]{4})(?P<month>[01][0-9](?P<day>[0-3][0-9]"
                        rf"(T?{time_pattern})?")
```
Note also that I have moved the timezone `Z` after the minute bracket since I think that just the hour with a timezone is also quite plausible.

Reply:
I fixed your proposal and adapt it re. line length limit